guacamole-ext/src/main/java/org/apache/guacamole/net/auth/DecoratingDirectory.java

```diff
@@ -0,0 +1,133 @@
+ *
+ * @param <ObjectType>
+ *     The type of objects stored within this Directory.
+ */
+public abstract class DecoratingDirectory<ObjectType extends Identifiable>
+        extends DelegatingDirectory<ObjectType> {
+
+    /**
+     * Creates a new DecoratingDirectory which decorates the objects within
+     * the given directory.
+     *
+     * @param directory
+     *     The Directory whose objects are being decorated.
+     */
+    public DecoratingDirectory(Directory<ObjectType> directory) {
+        super(directory);
+    }
+
+    /**
+     * Given an object retrieved from a Directory which originates from a
+     * different AuthenticationProvider, returns an identical type of object
+     * optionally wrapped with additional information, functionality, etc. If
+     * this directory chooses to decorate the object provided, it is up to the
+     * implementation of that decorated object to properly pass through
+     * operations as appropriate, as well as provide for an eventual
+     * undecorate() operation. All objects retrieved from this
+     * DecoratingDirectory will first be passed through this function.
+     *
+     * @param object
+     *     An object from a Directory which originates from a different
+     *     AuthenticationProvider.
+     *
+     * @return
+     *     An object which may have been decorated by this
+     *     DecoratingDirectory. If the object was not decorated, the original,
+     *     unmodified object may be returned instead.
+     *
+     * @throws GuacamoleException
+     *     If the provided object cannot be decorated due to an error.
+     */
+    protected abstract ObjectType decorate(ObjectType object)
+            throws GuacamoleException;
+
+    /**
+     * Given an object originally returned from a call to this
+     * DecoratingDirectory's decorate() function, reverses the decoration
+     * operation, returning the original object. This function is effectively
+     * the exact inverse of the decorate() function. The return value of
+     * undecorate(decorate(X)) must be identically X. All objects given to this
+     * DecoratingDirectory via add() or update() will first be passed through
+     * this function.
+     *
+     * @param object
+     *     An object which was originally returned by a call to this
+     *     DecoratingDirectory's decorate() function.
+     *
+     * @return
+     *     The original object which was provided to this DecoratingDirectory's
+     *     decorate() function.
+     *
+     * @throws GuacamoleException
+     *     If the provided object cannot be undecorated due to an error.
+     */
+    protected abstract ObjectType undecorate(ObjectType object)
+            throws GuacamoleException;
+
+    @Override
+    public ObjectType get(String identifier) throws GuacamoleException {
+
+        // Decorate only if object exists
+        ObjectType object = super.get(identifier);
+        if (object == null)
+            return object;
+
```

Comment:
This is more a curiosity question for me than anything, but is there some reason you chose this implementation rather than, for example:

    if (object != null)
        return decorate(object);

    return null;

?  I guess I'm a little fuzzy on why you'd return the object if there's nothing there, anyway, instead of flipping the logic?

Reply:
Yeah, that's a fair point. I'll change this to `return null`.